### DISCUSSION TITLE
Github Apps Token

### DISCUSSION THREAD

**Discussion by mirfanrafif** [community/community#24635] [Apps, API and Webhooks]
<p>Today, I am working on final assignment for graduation criteria in my university. I am making a system that can check students’ answer (test, parse and send it to REST API backend). And I make a system that can generate repository by it’s template. I am using a Github Apps to interact with Github API.<br>
But I have a problem when creating Github Apps installations access token. Sometimes it goes wrong.  It respond me like this :</p>
<p>Client error: <code>POST https://api.github.com/app/installations/INSTALLATIONS_ID/access_tokens</code> resulted in a <code>401 Unauthorized</code> response:\n{“message”:"‘Expiration time’ claim (‘exp’) is too far in the future",“documentation_url”:“<a href="https://docs.github.com/rest" class="inline-onebox" rel="noopener nofollow ugc">GitHub REST API - GitHub Docs</a>” (truncated…)\n</p>
<p>Its started today. Yesterday and 1-2 month ago, I made this system (I mean block of code) and always working. Is there any new limitation to generate token every 1 minute and no shorter time?</p>

**Comment by mirfanrafif** [comment]
<p>I send my code here :</p>
<pre><code> private function createJwtToken()
{

    $configuration = Configuration::forSymmetricSigner(

        // You may use any HMAC variations (256, 384, and 512)

        new Signer\Rsa\Sha256(),

        // replace the value below with a key of your own!

        InMemory::plainText(env('GITHUB_APPS_SECRET_KEY', ''))

        // You may also override the JOSE encoder/decoder if needed by providing extra arguments here

    );

    $now   = new DateTimeImmutable();

    $token = $configuration-&gt;builder(new MyClaimsFormatter())

        -&gt;issuedBy(env('GITHUB_APPS_ID', 0))

        -&gt;issuedAt($now)

        -&gt;expiresAt($now-&gt;modify('+10 minute'))

        -&gt;getToken($configuration-&gt;signer(), $configuration-&gt;signingKey());

    info($token-&gt;toString());

    return $token-&gt;toString();

}

private function getAccessToken(Client $http, $jwtToken)

{

    //get access token

    return $http-&gt;request('POST', 'app/installations/INSTALLATIONS_ID/access_tokens', [

        'headers' =&gt; [

            'Accept' =&gt; 'application/json',

            'Authorization' =&gt; 'Bearer ' . $jwtToken

        ]

    ]);

}
</code></pre>
<p>Like to see your answers😁</p>

**Comment by Katsute** [comment, marked as answer]
<aside class="quote quote-modified" data-post="2" data-topic="13830">
  <div class="title">
    <div class="quote-controls"></div>
    
    <a href="https://github.community/t/expiration-time-claim-exp-is-too-far-in-the-future-when-creating-an-access-token/13830/2">"'Expiration time' claim ('exp') is too far in the future" when creating an access token</a> <a class="badge-wrapper  bullet" href="/c/github-ecosystem/37"><span class="badge-category-bg" style="background-color: #28a745;"></span><span style="" data-drop-close="true" class="badge-category clear-badge" title="Discussion and support using GitHub’s REST, and GraphQL APIs, building Applications and OAuth Apps, the GitHub Marketplace, and third party integrations.">GitHub Ecosystem</span></a>
  </div>
  <blockquote>
    Okay so the documentation isn’t very clear, but maybe <a href="https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/#authenticating-as-a-github-app" rel="noopener nofollow ugc">this section on authenticating with JWT </a>is implying that the ‘exp’ can be at most 10 minutes later than the ‘iat’? 
If this is the case, then I think that 


The documentation needs to be more clear that 10 minutes is the maximum time hardcoded into the GitHub server (I’ve been misinterpreting that section to mean “in this example, 10 minutes is the maximum amount of time this token will be available for”) 


The GitHub server should be lenie…
  </blockquote>
</aside>

**Comment by mirfanrafif** [comment]
<p>thanks bro. thats really helpful</p>

**Comment by Sheriffmustapha** [comment]
import requests

# Your GitHub Personal Access Token
GITHUB_TOKEN = "your_github_personal_access_token"
USERNAME = "Sheriffmustapha"

# Repository details
repo_name = "30-Days-Python-Challenges"
repo_description = "A repository for 30 days of Python coding challenges."

# GitHub API URL for creating a repository
url = f"https://api.github.com/user/repos"

# Repository payload
payload = {
    "name": repo_name,
    "description": repo_description,
    "private": False,  # Set to True if you want the repo to be private
    "has_issues": True,
    "has_projects": True,
    "has_wiki": True
}

# Headers for the API request
headers = {
    "Authorization": f"token {GITHUB_TOKEN}",
    "Accept": "application/vnd.github.v3+json"
}

# Create the repository
response = requests.post(url, json=payload, headers=headers)

if response.status_code == 201:
    print(f"Repository '{repo_name}' created successfully!")
    print(f"Repo URL: https://github.com/{USERNAME}/{repo_name}")
else:
    print("Failed to create repository")
    print("Response:", response.json())
